independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -152,50 +153,35 @@ public ProtectionDomain getProtectionDomain() {
     }
 
     private JarFile readLockAcquireAndGetJarReference() {
-        while (true) {
-            readLock.lock();
-            final JarFile zipFileLocal = this.zipFile;
-            if (zipFileLocal != null) {
-                //Expected fast path: returns a reference to the open JarFile while owning the readLock
-                return zipFileLocal;
-            } else {
-                //This Lock implementation doesn't allow upgrading a readLock to a writeLock, so release it
-                //as we're going to need the WriteLock.
-                readLock.unlock();
-                //trigger the JarFile being (re)opened.
-                ensureJarFileIsOpen();
-                //Now since we no longer own any lock, we need to try again to obtain the readLock
-                //and check for the reference still being valid.
-                //This exposes us to a race with closing the just-opened JarFile;
-                //however this should be extremely rare, so we can trust we won't loop much;
-                //A counter doesn't seem necessary, as in fact we know that methods close()
-                //and resetInternalCaches() are invoked each at most once, which limits the amount
-                //of loops here in practice.
+        while (zipFileLoadingState.get() != ZipFileLoadingState.LOADED) {
+            if (zipFileLoadingState.compareAndSet(ZipFileLoadingState.NOT_LOADED, ZipFileLoadingState.LOADING)) {
+                return loadZipFile();
             }
+            Thread.yield();
```

Comment:
This still blocks, it just doesn't look like blocking (the thread can loop for an arbitrarily long time in this case, keeping the thread pinned the whole time because the pinning native call 
is still on the stack). I suspect that the only way around this is to not publish the class loader object at all, until the zip file is actually opened.

Reply:
That was my doubt too. I can simply remove this statement (and the other yield one on the close), but I'm just afraid that this could make the potential monopolization problem even worse. This implementation however should prevent the deadlock problem reported in the original issue (and that in all honesty I've never been able to reproduce anyway).

Reply:
That's true but the livelock could be especially bad with virtual threads since they are not preemptive. Unfortunately the only way to make them preemptive at that point is performing a blocking operation which is exactly the thing that we want to avoid. In a way or another this is probably still a sub-optimal solution, but I think better than the existing one. I'm not strongly opinionated in keeping or removing the `yield`s, please advice on what I should do about them.

Reply:
I think you have to keep them. Better would be to park/unpark but I don't see how that would be feasible, since the one part of the problem that this solves is avoiding a dependency on the owner thread to release waiter thread(s).